### PR TITLE
Use cert_reqs=CERT_REQUIRED by default

### LINKED ISSUE #1454: Shouldn't urllib3 verify SSL/TLS certificates by default ?
The README says:

> ... urllib3 brings many critical features that are missing from the Python standard libraries: 
> ...
> Client-side SSL/TLS verification.
> ...

and only in https://urllib3.readthedocs.io/en/latest/user-guide.html#certificate-verification it's explained that 

> By default, urllib3 does not verify HTTPS requests.

Because since [pep-0476](https://www.python.org/dev/peps/pep-0476) (ie python >= 2.7.9 or > 3.4 ) standard libraries verifies SSL certificates by default, nowadays projects switching from urllib2 or httplib to urllib3 without explicitly enabling certificate validation would loose the feature of certificate validation that was provided by standard library.

Of course, this would be an incompatible breaking change, like it was for python at the time, but is it really still a sane default to not verify certificates nowadays ?

### PR DESCRIPTION
**Pull request by sethmlarson**
Initial implementation of requiring and validating certificates when using HTTPS by default.
I'm sure I'll have to rewrite some tests, will work on that after I see all the failures. :)

The `InsecureRequestWarning` will still be emitted when explicitly using `CERT_NONE` or `CERT_OPTIONAL`.

Closes #1454.

### CODE REVIEW

**Comment by shazow** [comment]
Yay excited about this one. :)

Should we bolt on more docs to cert failure exceptions so people know what to do?

**Comment by sigmavirus24** [comment]
> Should we bolt on more docs to cert failure exceptions so people know what to do?

Only if those docs are actually helpful

**Comment by sethmlarson** [comment]
Definitely! I'll take a look at the docs as well tonight.

File: docs/user-guide.rst [hunk 1]
```diff
@@ -202,8 +202,13 @@ recommended to set the ``Content-Type`` header::
 Certificate verification
 ------------------------
 
+ .. note:: *New in version 1.25*
+
+    HTTPS connections are now verified by default
+    (``cert_reqs = 'CERT_REQUIRED'``) regardless of whether
+    CA certs are specified.
+
 It is highly recommended to always use SSL certificate verification.
```
[the remark this answers is not in the document]
Comment (theacodes) [reply docs/user-guide.rst]: Let's reword it, maybe "While you can disable certification verification, it is highly recommend to leave it on."

File: docs/user-guide.rst [hunk 2]
```diff
@@ -202,8 +202,13 @@ recommended to set the ``Content-Type`` header::
...
-**By default, urllib3 does not verify HTTPS requests**.
 
 In order to enable verification you will need a set of root certificates. The easiest
```
Comment (theacodes) [inline docs/user-guide.rst]: Is this still true?

Comment (sethmlarson) [reply docs/user-guide.rst]: I don't know if this was ever completely true? You could set `cert_reqs` to `REQUIRED` and we'd load the default verification locations.

So now by default urllib3 tries to load the default certificate locations via [`SSLContext.load_default_certs()`](https://docs.python.org/3.7/library/ssl.html#ssl.SSLContext.load_default_certs) which apparently works well on Windows and on non-Windows uses [`SSLContext.set_default_verify_locations()`](https://docs.python.org/3.7/library/ssl.html#ssl.SSLContext.set_default_verify_paths) which is also reliable in my experience.

The only platforms that will have to use certifi or their own bundle are platforms that don't have `SSLContext.load_default_certs()` or don't define `SSLContext` at all because our implementation of `SSLContext` doesn't define `load_default_certs()` (Should our implementation try to implement this?).

As for `contrib/`: `PyOpenSSLContext` defines a functioning `load_default_certs()` and `SecureTransportContext` defines the method but does absolutely nothing with it so maybe it will fail cert verification unless SecureTransport loads certs by default? Will have to look at docs.

To sum it up, it's a mixed bag of whether we load system certs if no certificate locations are given. :)

File: docs/user-guide.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (sethmlarson) [reply docs/user-guide.rst]: Done!

File: docs/user-guide.rst [hunk 3]
```diff
@@ -202,8 +202,13 @@ recommended to set the ``Content-Type`` header::
 Certificate verification
 ------------------------
 
+ .. note:: *New in version 1.25*
+
+    HTTPS connections are now verified by default
```
[the remark this answers is not in the document]
Comment (sethmlarson) [reply docs/user-guide.rst]: Done!

**Comment by sethmlarson** [comment]
Thanks for the comments, I've updated the documentation in the areas mentioned. What do we think about the wordings now?

File: docs/user-guide.rst (hunk 2 above)
Comment (theacodes) [reply docs/user-guide.rst]: Then we should probs remove this, yeah?

Comment (sethmlarson) [reply docs/user-guide.rst]: When you say this you mean the whole section on certifi? I'm not sure what your comment is directed at. If we remove that subsection then the section as a whole is pretty bare and the last part mentions system certificates as well. Maybe this whole section needs a rewrite?

File: docs/user-guide.rst [hunk 4]
```diff
@@ -202,11 +202,15 @@ recommended to set the ``Content-Type`` header::
...
-**By default, urllib3 does not verify HTTPS requests**.
+ .. note:: *New in version 1.25*
 
-In order to enable verification you will need a set of root certificates. The easiest
-and most reliable method is to use the `certifi <https://certifi.io/>`_ package which provides Mozilla's root certificate bundle::
+    HTTPS connections are now verified by default (``cert_reqs = 'CERT_REQUIRED'``).
+
+While you can disable certification verification, it is highly recommend to leave it on.
+
+For certificate verification to work properly you will need a set of root certificates. Unless otherwise
```
Comment (shazow) [inline docs/user-guide.rst]: > For certificate verification to work properly you will need a set of root certificates. Unless otherwise specified urllib3 will try to load the default system certificate stores.

Aren't these two sentences contradictory? The first one sounds like you have to do something additional for urllib3 to work at all, then the second sentence sounds like it does something sensible by default and you don't need to do anything.

IMO remove the first sentence, or rephrase.

Comment (sethmlarson) [reply docs/user-guide.rst]: Sounds good to me :)

**Comment by shazow** [comment]
:custard:

**Comment by sethmlarson** [comment]
We can change the documentation in a separate PR, would like to merge this change into the TLS 1.3 PR so we can finally land that monster.
